packages/machine/test/integration/protocols.spec.ts

```diff
@@ -0,0 +1,215 @@
+    const [key] = [...appInstances.keys()].filter(key => {
+      return (
+        key !==
+        mininodeA.scm.get(AddressZero)!.toJson().freeBalanceAppIndexes[0][1]
+      );
+    });
+
+    // increments comes back as 0 and free balance is not decremented
+
+    await mininodeA.ie.runUninstallProtocol(mininodeA.scm, {
+      appIdentityHash: key,
+      initiatingXpub: mininodeA.xpub,
+      respondingXpub: mininodeB.xpub,
+      multisigAddress: AddressZero
+    });
+
+    mr.assertNoPending();
+
+    const addressOne = "0x0000000000000000000000000000000000000001";
+
+    mininodeB.scm.set(
+      addressOne,
+      (await mininodeB.ie.runSetupProtocol({
+        initiatingXpub: mininodeB.xpub,
+        respondingXpub: mininodeC.xpub,
+        multisigAddress: "0x0000000000000000000000000000000000000001"
+      })).get(addressOne)!
+    );
+
+    mr.assertNoPending();
+
+    expect(mininodeA.scm.size).toBe(1);
+    expect(mininodeB.scm.size).toBe(2);
+    expect(mininodeC.scm.size).toBe(1);
+
+    await mininodeA.ie.runInstallVirtualAppProtocol(mininodeA.scm, {
+      initiatingXpub: mininodeA.xpub,
+      intermediaryXpub: mininodeB.xpub,
+      respondingXpub: mininodeC.xpub,
+      defaultTimeout: 100,
+      appInterface: {
+        addr: appDefinition.address,
+        stateEncoding:
+          "tuple(address player1, address player2, uint256 counter)",
+        actionEncoding: "tuple(uint256 increment)"
+      },
+      initialState: {
+        player1: AddressZero,
+        player2: AddressZero,
+        counter: 0
+      },
+      initiatingBalanceDecrement: bigNumberify(0),
+      respondingBalanceDecrement: bigNumberify(0)
+    });
+
+    expect(mininodeA.scm.size).toBe(2);
+
+    const [virtualKey] = [...mininodeA.scm.keys()].filter(key => {
+      return key !== AddressZero;
+    });
+
+    const [appInstance] = [
+      ...mininodeA.scm.get(virtualKey)!.appInstances.values()
+    ];
+
+    expect(appInstance.isVirtualApp);
+
+    await mininodeA.ie.runUpdateProtocol(mininodeA.scm, {
+      initiatingXpub: mininodeA.xpub,
+      respondingXpub: mininodeC.xpub,
+      multisigAddress: virtualKey,
+      appIdentityHash: appInstance.identityHash,
+      newState: {
+        player1: AddressZero,
+        player2: AddressZero,
+        counter: 1
+      }
+    });
+
+    await mininodeA.ie.runTakeActionProtocol(mininodeA.scm, {
+      initiatingXpub: mininodeA.xpub,
+      respondingXpub: mininodeC.xpub,
+      multisigAddress: virtualKey,
+      appIdentityHash: appInstance.identityHash,
+      action: {
+        increment: 1
+      }
+    });
+
+    await mininodeA.ie.runUninstallVirtualAppProtocol(mininodeA.scm, {
+      initiatingXpub: mininodeA.xpub,
+      intermediaryXpub: mininodeB.xpub,
+      respondingXpub: mininodeC.xpub,
+      targetAppIdentityHash: appInstance.identityHash,
+      targetAppState: {
+        player1: AddressZero,
+        player2: AddressZero,
+        counter: 1
+      }
+    });
+  });
+});
```

Comment:
Can you expand this test suite so there is one test suite per protocol and ideally multiple tests per protocol for different scenarios e.g., bad input, invalid state, etc.